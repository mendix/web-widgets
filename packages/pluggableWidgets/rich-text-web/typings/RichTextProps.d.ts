/**
 * This file was generated from RichText.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { ActionValue, DynamicValue, EditableValue } from "mendix";

export type MenubarModeEnum = "hide" | "basic" | "full" | "custom";

export type PresetEnum = "basic" | "standard" | "full" | "custom";

export type ToolbarModeEnum = "sliding" | "floating" | "scrolling" | "wrap";

export type ToolbarLocationEnum = "auto" | "top" | "bottom" | "inline";

export type ContextmenutypeEnum = "native" | "richtext";

export type ReadOnlyStyleEnum = "text" | "bordered" | "readPanel";

export type WidthUnitEnum = "percentage" | "pixels";

export type HeightUnitEnum = "percentageOfWidth" | "pixels" | "percentageOfParent";

export type ResizeEnum = "false" | "true" | "both";

export type OnChangeTypeEnum = "onLeave" | "onDataChange";

export type ToolbarConfigEnum = "basic" | "advanced";

export type CtItemTypeEnum = "separator" | "aligncenter" | "alignjustify" | "alignleft" | "alignnone" | "alignright" | "blockquote" | "backcolor" | "blocks" | "bold" | "copy" | "cut" | "fontfamily" | "fontsize" | "forecolor" | "hr" | "indent" | "italic" | "lineheight" | "newdocument" | "outdent" | "paste" | "pastetext" | "print" | "redo" | "remove" | "removeformat" | "selectall" | "strikethrough" | "subscript" | "superscript" | "underline" | "undo" | "visualaid" | "accordion" | "code" | "anchor" | "charmap" | "codesample" | "ltr" | "rtl" | "emoticons" | "fullscreen" | "help" | "image" | "insertdatetime" | "link" | "openlink" | "unlink" | "bullist" | "numlist" | "media" | "pagebreak" | "preview" | "searchreplace" | "table" | "tabledelete" | "tableinsertdialog" | "visualblocks" | "visualchars" | "wordcount";

export interface AdvancedConfigType {
    ctItemType: CtItemTypeEnum;
}

export type MenubarConfigEnum = "basic" | "advanced";

export type MenubarItemTypeEnum = "file" | "edit" | "insert" | "view" | "format" | "table" | "tools" | "help";

export interface AdvancedMenubarConfigType {
    menubarItemType: MenubarItemTypeEnum;
}

export interface AdvancedConfigPreviewType {
    ctItemType: CtItemTypeEnum;
}

export interface AdvancedMenubarConfigPreviewType {
    menubarItemType: MenubarItemTypeEnum;
}

export interface RichTextContainerProps {
    name: string;
    tabIndex?: number;
    id: string;
    stringAttribute: EditableValue<string>;
    menubarMode: MenubarModeEnum;
    enableStatusBar: boolean;
    preset: PresetEnum;
    toolbarMode: ToolbarModeEnum;
    toolbarLocation: ToolbarLocationEnum;
    quickbars: boolean;
    contextmenutype: ContextmenutypeEnum;
    readOnlyStyle: ReadOnlyStyleEnum;
    widthUnit: WidthUnitEnum;
    width: number;
    heightUnit: HeightUnitEnum;
    height: number;
    minHeight: number;
    resize: ResizeEnum;
    onChange?: ActionValue;
    onFocus?: ActionValue;
    onBlur?: ActionValue;
    onKeyPress?: ActionValue;
    onChangeType: OnChangeTypeEnum;
    extended_valid_elements?: DynamicValue<string>;
    spellCheck: boolean;
    highlight_on_focus: boolean;
    content_css?: DynamicValue<string>;
    toolbarConfig: ToolbarConfigEnum;
    basicstyle: boolean;
    extendedstyle: boolean;
    textalign: boolean;
    clipboard: boolean;
    fontstyle: boolean;
    paragraph: boolean;
    document: boolean;
    history: boolean;
    accordion: boolean;
    code: boolean;
    anchor: boolean;
    direction: boolean;
    link: boolean;
    list: boolean;
    preview: boolean;
    table: boolean;
    visualaid: boolean;
    media: boolean;
    util: boolean;
    emoticon: boolean;
    remove: boolean;
    advancedConfig: AdvancedConfigType[];
    menubarConfig: MenubarConfigEnum;
    fileMenubar: boolean;
    editMenubar: boolean;
    insertMenubar: boolean;
    viewMenubar: boolean;
    formatMenubar: boolean;
    tableMenubar: boolean;
    toolsMenubar: boolean;
    helpMenubar: boolean;
    advancedMenubarConfig: AdvancedMenubarConfigType[];
}

export interface RichTextPreviewProps {
    readOnly: boolean;
    stringAttribute: string;
    menubarMode: MenubarModeEnum;
    enableStatusBar: boolean;
    preset: PresetEnum;
    toolbarMode: ToolbarModeEnum;
    toolbarLocation: ToolbarLocationEnum;
    quickbars: boolean;
    contextmenutype: ContextmenutypeEnum;
    readOnlyStyle: ReadOnlyStyleEnum;
    widthUnit: WidthUnitEnum;
    width: number | null;
    heightUnit: HeightUnitEnum;
    height: number | null;
    minHeight: number | null;
    resize: ResizeEnum;
    onChange: {} | null;
    onFocus: {} | null;
    onBlur: {} | null;
    onKeyPress: {} | null;
    onChangeType: OnChangeTypeEnum;
    extended_valid_elements: string;
    spellCheck: boolean;
    highlight_on_focus: boolean;
    content_css: string;
    toolbarConfig: ToolbarConfigEnum;
    basicstyle: boolean;
    extendedstyle: boolean;
    textalign: boolean;
    clipboard: boolean;
    fontstyle: boolean;
    paragraph: boolean;
    document: boolean;
    history: boolean;
    accordion: boolean;
    code: boolean;
    anchor: boolean;
    direction: boolean;
    link: boolean;
    list: boolean;
    preview: boolean;
    table: boolean;
    visualaid: boolean;
    media: boolean;
    util: boolean;
    emoticon: boolean;
    remove: boolean;
    advancedConfig: AdvancedConfigPreviewType[];
    menubarConfig: MenubarConfigEnum;
    fileMenubar: boolean;
    editMenubar: boolean;
    insertMenubar: boolean;
    viewMenubar: boolean;
    formatMenubar: boolean;
    tableMenubar: boolean;
    toolsMenubar: boolean;
    helpMenubar: boolean;
    advancedMenubarConfig: AdvancedMenubarConfigPreviewType[];
}
