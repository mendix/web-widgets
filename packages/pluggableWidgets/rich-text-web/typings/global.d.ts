export interface MXGlobalObject {
    remoteUrl: string;
}

declare global {
    interface Window {
        mx: MXGlobalObject;
    }
}
