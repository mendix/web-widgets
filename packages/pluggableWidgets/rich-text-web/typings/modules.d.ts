declare module "*.css";
declare module "*.scss";

// Add quill-resize-module declaration
declare module "quill-resize-module" {
    import Quill from "quill";

    interface ResizeModuleOptions {
        modules?: string[];
        handleStyles?: {
            backgroundColor?: string;
            border?: string;
            boxSizing?: string;
        };
        displayStyles?: {
            backgroundColor?: string;
            border?: string;
            color?: string;
        };
        toolbarStyles?: {
            backgroundColor?: string;
            border?: string;
            color?: string;
            boxShadow?: string;
        };
        overlayStyles?: {
            border?: string;
            boxSizing?: string;
        };
        embedTags?: string[];
        tools?: Array<
            | string
            | {
                  text: string;
                  verify: (activeEle: HTMLElement) => boolean;
                  handler: (evt: MouseEvent, button: HTMLElement, activeEle: HTMLElement) => void;
              }
        >;
        locale?: {
            altTip?: string;
            inputTip?: string;
            floatLeft?: string;
            floatRight?: string;
            center?: string;
            restore?: string;
        };
    }

    interface ResizeModuleConstructor {
        new (quill: Quill, options: ResizeModuleOptions): any;
        Modules?: any;
    }

    const ResizeModule: ResizeModuleConstructor;
    export default ResizeModule;
}

// Add the dist/resize module declaration
declare module "quill-resize-module/dist/resize" {
    export * from "quill-resize-module";
    export { default } from "quill-resize-module";
}
