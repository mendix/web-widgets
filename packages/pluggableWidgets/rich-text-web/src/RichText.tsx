import { ValidationAlert } from "@mendix/widget-plugin-component-kit/Alert";
import { getDimensions } from "@mendix/widget-plugin-platform/utils/get-dimensions";
import classNames from "classnames";
import { createElement, Fragment } from "react";
import { RichTextContainerProps } from "../typings/RichTextProps";
import BundledEditor from "./components/Editor";
import "./ui/RichText.scss";
import { constructWrapperStyle } from "./utils/helpers";
import { createMenubar } from "./utils/menubar";
import { createPreset } from "./utils/presets";

export default function RichText(props: RichTextContainerProps): JSX.Element {
    const {
        stringAttribute,
        id,
        width: w,
        height: h,
        widthUnit,
        heightUnit,
        preset,
        menubarMode,
        readOnlyStyle,
        enableStatusBar,
        resize
    } = props;

    const { width, height } = getDimensions({
        width: w,
        widthUnit,
        height: h,
        heightUnit
    });
    const wrapperAttributes = stringAttribute?.readOnly && readOnlyStyle !== "readPanel" ? { readOnly: true } : {};

    const presets = createPreset(preset, props);
    const menubar = createMenubar(menubarMode, props);
    const wrapperStyle = constructWrapperStyle(props, { width, height });
    return (
        <Fragment>
            <div
                id={id}
                className={classNames(
                    "widget-rich-text",
                    `${stringAttribute?.readOnly ? `editor-${readOnlyStyle}` : ""}`,
                    {
                        "form-control": props.toolbarLocation === "inline",
                        "widget-rich-text-min-height": heightUnit !== "pixels" && !stringAttribute?.readOnly,
                        "widget-rich-text-min-height-readonly": heightUnit !== "pixels" && stringAttribute?.readOnly
                    }
                )}
                style={wrapperStyle}
                {...wrapperAttributes}
            >
                {stringAttribute.status === "loading" || stringAttribute.status !== "available" ? (
                    <div className="mx-progress"></div>
                ) : (
                    <BundledEditor
                        {...props}
                        menubar={menubar}
                        toolbar={presets.toolbar}
                        editorHeight={height}
                        editorWidth={width}
                        key={`${String(stringAttribute.readOnly)}_${id}_${props.content_css?.value}`}
                        resize={enableStatusBar ? resize : "false"}
                    />
                )}
            </div>
            <ValidationAlert>{stringAttribute.validation}</ValidationAlert>
        </Fragment>
    );
}
