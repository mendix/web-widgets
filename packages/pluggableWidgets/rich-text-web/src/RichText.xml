<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.custom.richtext.RichText" needsEntityContext="true" pluginWidget="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
    <name>Rich Text</name>
    <description>Rich inline or toolbar text editing</description>
    <studioProCategory>Input elements</studioProCategory>
    <studioCategory>Input Elements</studioCategory>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data source">
                <property key="stringAttribute" type="attribute">
                    <caption>Value attribute</caption>
                    <description>The attribute used for the content of the text editor, recommendation is to use an unlimited string data type.</description>
                    <attributeTypes>
                        <attributeType name="String" />
                    </attributeTypes>
                </property>
            </propertyGroup>
            <propertyGroup caption="General">
                <systemProperty key="Label" />
                <property key="menubarMode" type="enumeration" defaultValue="hide">
                    <caption>Menu bar</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="hide">Hidden</enumerationValue>
                        <enumerationValue key="basic">Basic</enumerationValue>
                        <enumerationValue key="full">Full</enumerationValue>
                        <enumerationValue key="custom">Custom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="enableStatusBar" type="boolean" defaultValue="true">
                    <caption>Enable status bar</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Toolbar">
                <property key="preset" type="enumeration" defaultValue="basic">
                    <caption>Toolbar</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="basic">Basic</enumerationValue>
                        <enumerationValue key="standard">Standard</enumerationValue>
                        <enumerationValue key="full">Full</enumerationValue>
                        <enumerationValue key="custom">Custom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="toolbarMode" type="enumeration" defaultValue="sliding">
                    <caption>Mode</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="sliding">Sliding</enumerationValue>
                        <enumerationValue key="floating">Floating</enumerationValue>
                        <enumerationValue key="scrolling">Scrolling</enumerationValue>
                        <enumerationValue key="wrap">Wrap</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="toolbarLocation" type="enumeration" defaultValue="auto">
                    <caption>Location</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="auto">Auto</enumerationValue>
                        <enumerationValue key="top">Top</enumerationValue>
                        <enumerationValue key="bottom">Bottom</enumerationValue>
                        <enumerationValue key="inline">Inline</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="quickbars" type="boolean" defaultValue="false">
                    <caption>Enable quick toolbar</caption>
                    <description />
                </property>
                <property key="contextmenutype" type="enumeration" defaultValue="native">
                    <caption>Context menu</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="native">Native</enumerationValue>
                        <enumerationValue key="richtext">Rich Text</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Editability">
                <systemProperty key="Editability" />
                <property key="readOnlyStyle" type="enumeration" defaultValue="text">
                    <caption>Read-only style</caption>
                    <description>How the rich text editor will appear in read-only mode.</description>
                    <enumerationValues>
                        <enumerationValue key="text">Rich text</enumerationValue>
                        <enumerationValue key="bordered">Bordered</enumerationValue>
                        <enumerationValue key="readPanel">Read panel</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Visibility">
                <systemProperty key="Visibility" />
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Dimensions">
            <propertyGroup caption="Dimensions">
                <property key="widthUnit" type="enumeration" defaultValue="percentage">
                    <caption>Width unit</caption>
                    <description>Percentage: portion of parent size. Pixels: absolute amount of pixels.</description>
                    <enumerationValues>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="width" type="integer" defaultValue="100">
                    <caption>Width</caption>
                    <description />
                </property>
                <property key="heightUnit" type="enumeration" defaultValue="pixels">
                    <caption>Height unit</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="percentageOfWidth">Percentage of width</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentageOfParent">Percentage of parent</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="height" type="integer" defaultValue="250">
                    <caption>Height</caption>
                    <description />
                </property>
                <property key="minHeight" type="integer" defaultValue="250">
                    <caption>Minimum height</caption>
                    <description>Editor's minimum height. The number value is in pixels.</description>
                </property>
                <property key="resize" type="enumeration" defaultValue="false">
                    <caption>Editor resize</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="false">No</enumerationValue>
                        <enumerationValue key="true">Vertical</enumerationValue>
                        <enumerationValue key="both">Both</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Events">
            <propertyGroup caption="Events">
                <property key="onChange" type="action">
                    <caption>On change</caption>
                    <description />
                </property>
                <property key="onFocus" type="action">
                    <caption>On enter</caption>
                    <description />
                </property>
                <property key="onBlur" type="action">
                    <caption>On leave</caption>
                    <description />
                </property>
                <property key="onKeyPress" type="action">
                    <caption>On key press</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="On change behavior">
                <property key="onChangeType" type="enumeration" defaultValue="onLeave">
                    <caption>Editor resize</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="onLeave">When user leaves input field</enumerationValue>
                        <enumerationValue key="onDataChange">While user is entering data</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Advanced">
            <propertyGroup caption="Advanced">
                <property key="extended_valid_elements" type="textTemplate" required="false">
                    <caption>Additional valid elements</caption>
                    <description>Additional sanitization rules to allow certain tag, sample: script[src|async]</description>
                </property>
                <property key="spellCheck" type="boolean" defaultValue="false">
                    <caption>Enable spell checking</caption>
                    <description />
                </property>
                <property key="highlight_on_focus" type="boolean" defaultValue="false">
                    <caption>Highlight on focus</caption>
                    <description />
                </property>
                <property key="content_css" type="textTemplate" required="false">
                    <caption>Additional stylesheet</caption>
                    <description>Additional stylesheet to be included in the rich text content</description>
                </property>
            </propertyGroup>
        </propertyGroup>

        <propertyGroup caption="Custom toolbar">
            <propertyGroup caption="Custom toolbar">
                <property key="toolbarConfig" type="enumeration" defaultValue="basic">
                    <caption>Toolbar group</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="basic">Basic</enumerationValue>
                        <enumerationValue key="advanced">Advanced</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="basicstyle" type="boolean" defaultValue="true">
                    <caption>Basic style</caption>
                    <description />
                </property>
                <property key="extendedstyle" type="boolean" defaultValue="true">
                    <caption>Extended style</caption>
                    <description />
                </property>
                <property key="textalign" type="boolean" defaultValue="true">
                    <caption>Text alignment</caption>
                    <description />
                </property>
                <property key="clipboard" type="boolean" defaultValue="true">
                    <caption>Clipboard</caption>
                    <description />
                </property>
                <property key="fontstyle" type="boolean" defaultValue="true">
                    <caption>Font style</caption>
                    <description />
                </property>
                <property key="paragraph" type="boolean" defaultValue="true">
                    <caption>Paragraph</caption>
                    <description />
                </property>
                <property key="document" type="boolean" defaultValue="true">
                    <caption>Document</caption>
                    <description />
                </property>
                <property key="history" type="boolean" defaultValue="true">
                    <caption>History</caption>
                    <description />
                </property>
                <property key="accordion" type="boolean" defaultValue="true">
                    <caption>Accordion</caption>
                    <description />
                </property>
                <property key="code" type="boolean" defaultValue="true">
                    <caption>Code</caption>
                    <description />
                </property>
                <property key="anchor" type="boolean" defaultValue="true">
                    <caption>Anchor</caption>
                    <description />
                </property>
                <property key="direction" type="boolean" defaultValue="true">
                    <caption>Text direction</caption>
                    <description />
                </property>
                <property key="link" type="boolean" defaultValue="true">
                    <caption>Link</caption>
                    <description />
                </property>
                <property key="list" type="boolean" defaultValue="true">
                    <caption>List</caption>
                    <description />
                </property>
                <property key="preview" type="boolean" defaultValue="true">
                    <caption>Preview</caption>
                    <description />
                </property>
                <property key="table" type="boolean" defaultValue="true">
                    <caption>Table</caption>
                    <description />
                </property>
                <property key="visualaid" type="boolean" defaultValue="true">
                    <caption>Visual aid</caption>
                    <description />
                </property>
                <property key="media" type="boolean" defaultValue="true">
                    <caption>Media</caption>
                    <description />
                </property>
                <property key="util" type="boolean" defaultValue="true">
                    <caption>Utility</caption>
                    <description />
                </property>
                <property key="emoticon" type="boolean" defaultValue="true">
                    <caption>Emoticon</caption>
                    <description />
                </property>
                <property key="remove" type="boolean" defaultValue="true">
                    <caption>Removal</caption>
                    <description />
                </property>
                <property key="advancedConfig" type="object" isList="true" required="false">
                    <caption>Advanced groups</caption>
                    <description />
                    <properties>
                        <property key="ctItemType" type="enumeration" required="true" defaultValue="separator">
                            <caption>Button</caption>
                            <category>Item</category>
                            <description>Button Type</description>
                            <enumerationValues>
                                <enumerationValue key="separator">separator</enumerationValue>
                                <enumerationValue key="aligncenter">aligncenter</enumerationValue>
                                <enumerationValue key="alignjustify">alignjustify</enumerationValue>
                                <enumerationValue key="alignleft">alignleft</enumerationValue>
                                <enumerationValue key="alignnone">alignnone</enumerationValue>
                                <enumerationValue key="alignright">alignright</enumerationValue>
                                <enumerationValue key="blockquote">blockquote</enumerationValue>
                                <enumerationValue key="backcolor">backcolor</enumerationValue>
                                <enumerationValue key="blocks">blocks</enumerationValue>
                                <enumerationValue key="bold">bold</enumerationValue>
                                <enumerationValue key="copy">copy</enumerationValue>
                                <enumerationValue key="cut">cut</enumerationValue>
                                <enumerationValue key="fontfamily">fontfamily</enumerationValue>
                                <enumerationValue key="fontsize">fontsize</enumerationValue>
                                <enumerationValue key="forecolor">forecolor</enumerationValue>
                                <enumerationValue key="hr">hr</enumerationValue>
                                <enumerationValue key="indent">indent</enumerationValue>
                                <enumerationValue key="italic">italic</enumerationValue>
                                <enumerationValue key="lineheight">lineheight</enumerationValue>
                                <enumerationValue key="newdocument">newdocument</enumerationValue>
                                <enumerationValue key="outdent">outdent</enumerationValue>
                                <enumerationValue key="paste">paste</enumerationValue>
                                <enumerationValue key="pastetext">pastetext</enumerationValue>
                                <enumerationValue key="print">print</enumerationValue>
                                <enumerationValue key="redo">redo</enumerationValue>
                                <enumerationValue key="remove">remove</enumerationValue>
                                <enumerationValue key="removeformat">removeformat</enumerationValue>
                                <enumerationValue key="selectall">selectall</enumerationValue>
                                <enumerationValue key="strikethrough">strikethrough</enumerationValue>
                                <enumerationValue key="subscript">subscript</enumerationValue>
                                <enumerationValue key="superscript">superscript</enumerationValue>
                                <enumerationValue key="underline">underline</enumerationValue>
                                <enumerationValue key="undo">undo</enumerationValue>
                                <enumerationValue key="visualaid">visualaid</enumerationValue>
                                <enumerationValue key="accordion">accordion</enumerationValue>
                                <enumerationValue key="code">code</enumerationValue>
                                <enumerationValue key="anchor">anchor</enumerationValue>
                                <enumerationValue key="charmap">charmap</enumerationValue>
                                <enumerationValue key="codesample">codesample</enumerationValue>
                                <enumerationValue key="ltr">ltr</enumerationValue>
                                <enumerationValue key="rtl">rtl</enumerationValue>
                                <enumerationValue key="emoticons">emoticons</enumerationValue>
                                <enumerationValue key="fullscreen">fullscreen</enumerationValue>
                                <enumerationValue key="help">help</enumerationValue>
                                <enumerationValue key="image">image</enumerationValue>
                                <enumerationValue key="insertdatetime">insertdatetime</enumerationValue>
                                <enumerationValue key="link">link</enumerationValue>
                                <enumerationValue key="openlink">openlink</enumerationValue>
                                <enumerationValue key="unlink">unlink</enumerationValue>
                                <enumerationValue key="bullist">bullist</enumerationValue>
                                <enumerationValue key="numlist">numlist</enumerationValue>
                                <enumerationValue key="media">media</enumerationValue>
                                <enumerationValue key="pagebreak">pagebreak</enumerationValue>
                                <enumerationValue key="preview">preview</enumerationValue>
                                <enumerationValue key="searchreplace">searchreplace</enumerationValue>
                                <enumerationValue key="table">table</enumerationValue>
                                <enumerationValue key="tabledelete">tabledelete</enumerationValue>
                                <enumerationValue key="tableinsertdialog">tableinsertdialog</enumerationValue>
                                <enumerationValue key="visualblocks">visualblocks</enumerationValue>
                                <enumerationValue key="visualchars">visualchars</enumerationValue>
                                <enumerationValue key="wordcount">wordcount</enumerationValue>
                            </enumerationValues>
                        </property>
                    </properties>
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Custom menubar">
            <propertyGroup caption="Custom menubar">
                <property key="menubarConfig" type="enumeration" defaultValue="basic">
                    <caption>Toolbar group</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="basic">Basic</enumerationValue>
                        <enumerationValue key="advanced">Advanced</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="fileMenubar" type="boolean" defaultValue="true">
                    <caption>File</caption>
                    <description />
                </property>
                <property key="editMenubar" type="boolean" defaultValue="true">
                    <caption>Edit</caption>
                    <description />
                </property>
                <property key="insertMenubar" type="boolean" defaultValue="true">
                    <caption>Insert</caption>
                    <description />
                </property>
                <property key="viewMenubar" type="boolean" defaultValue="true">
                    <caption>View</caption>
                    <description />
                </property>
                <property key="formatMenubar" type="boolean" defaultValue="true">
                    <caption>Format</caption>
                    <description />
                </property>
                <property key="tableMenubar" type="boolean" defaultValue="true">
                    <caption>Table</caption>
                    <description />
                </property>
                <property key="toolsMenubar" type="boolean" defaultValue="true">
                    <caption>Tools</caption>
                    <description />
                </property>
                <property key="helpMenubar" type="boolean" defaultValue="true">
                    <caption>Help</caption>
                    <description />
                </property>
                <property key="advancedMenubarConfig" type="object" isList="true" required="false">
                    <caption>Advanced groups</caption>
                    <description />
                    <properties>
                        <property key="menubarItemType" type="enumeration" required="true" defaultValue="file">
                            <caption>Button</caption>
                            <category>Item</category>
                            <description>Button Type</description>
                            <enumerationValues>
                                <enumerationValue key="file">File</enumerationValue>
                                <enumerationValue key="edit">Edit</enumerationValue>
                                <enumerationValue key="insert">Insert</enumerationValue>
                                <enumerationValue key="view">View</enumerationValue>
                                <enumerationValue key="format">Format</enumerationValue>
                                <enumerationValue key="table">Table</enumerationValue>
                                <enumerationValue key="tools">Tools</enumerationValue>
                                <enumerationValue key="help">Help</enumerationValue>
                            </enumerationValues>
                        </property>
                    </properties>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
