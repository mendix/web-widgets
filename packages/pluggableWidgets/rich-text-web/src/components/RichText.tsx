import { createElement, useState } from "react";
import { getDimensions } from "@mendix/pluggable-widgets-commons";
import { RichTextContainerProps } from "../../typings/RichTextProps";
import loadingCircleSvg from "../ui/loading-circle.svg";
import classNames from "classnames";
import { Editor } from "./Editor";

export function RichText(props: RichTextContainerProps): JSX.Element | null {
    const { width: w, height: h, widthUnit, heightUnit, readOnlyStyle, name, id } = props;
    const [element, setElement] = useState<HTMLElement | null>(null);

    if (props.stringAttribute.status !== "available") {
        return <img src={loadingCircleSvg} className="widget-rich-text-loading-spinner" alt="" aria-hidden />;
    }

    const { width, height } = getDimensions({
        width: w,
        widthUnit,
        height: h,
        heightUnit
    });

    const readOnly = props.stringAttribute.readOnly;

    return (
        <div
            id={id}
            className={classNames("widget-rich-text", `${readOnly ? `editor-${readOnlyStyle}` : ""}`)}
            style={{ width, height }}
        >
            <div ref={setElement} id={name} />
            {element ? <Editor element={element} widgetProps={props} /> : null}
        </div>
    );
}
