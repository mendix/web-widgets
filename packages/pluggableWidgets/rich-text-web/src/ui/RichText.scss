$rte-brand-primary: #264ae5;
.widget-rich-text {
    overflow-y: auto;
    display: flex;
    flex-direction: column;
    .tox-tinymce {
        border-radius: 4px;
        border-width: 1px;
        flex-grow: 1;
    }

    .tox .tox-edit-area::before {
        border-color: $rte-brand-primary;
        border-radius: 1px;
        border-width: 1px;
    }

    &.editor-bordered,
    &.editor-text {
        .tox-edit-area__iframe {
            background-color: transparent;
        }
    }

    &.editor-text {
        border: unset;

        .tox-tinymce {
            border: unset;
        }
    }

    .mce-content-body {
        width: 100%;
        outline: 0;
    }
}

.modal-dialog {
    & ~ .tox-tinymce-inline {
        &.tox-tinymce--toolbar-sticky-off {
            z-index: 102;
        }
    }
}
