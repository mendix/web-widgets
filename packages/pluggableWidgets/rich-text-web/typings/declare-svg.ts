declare module "*.svg" {
    const content: string;
    export = content;
}

declare module 'quill-better-table';

declare module '*.css';