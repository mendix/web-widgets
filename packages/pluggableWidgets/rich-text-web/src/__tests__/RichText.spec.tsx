import { EditableValueBuilder } from "@mendix/widget-plugin-test-utils";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import { createElement } from "react";
import { RichTextContainerProps } from "../../typings/RichTextProps";

Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn()
    }))
});

import RichText from "../RichText";

describe("Rich Text", () => {
    let defaultProps: RichTextContainerProps;
    beforeEach(() => {
        defaultProps = {
            name: "RichText",
            id: "RichText1",
            stringAttribute: new EditableValueBuilder<string>().withValue("Rich text default value").build(),
            menubarMode: "basic",
            enableStatusBar: true,
            preset: "basic",
            toolbarMode: "sliding",
            toolbarLocation: "bottom",
            widthUnit: "percentage",
            width: 100,
            heightUnit: "percentageOfWidth",
            height: 75,
            minHeight: 75,
            spellCheck: false,
            highlight_on_focus: false,
            resize: "false",
            toolbarConfig: "basic",
            basicstyle: true,
            extendedstyle: true,
            textalign: true,
            clipboard: true,
            fontstyle: true,
            paragraph: true,
            document: true,
            history: true,
            accordion: true,
            code: true,
            anchor: true,
            direction: true,
            link: true,
            list: true,
            preview: true,
            table: true,
            visualaid: true,
            media: true,
            util: true,
            emoticon: true,
            remove: true,
            advancedConfig: [],
            menubarConfig: "basic",
            fileMenubar: true,
            editMenubar: true,
            insertMenubar: true,
            viewMenubar: true,
            formatMenubar: true,
            tableMenubar: true,
            toolsMenubar: true,
            helpMenubar: true,
            quickbars: true,
            readOnlyStyle: "text",
            advancedMenubarConfig: [],
            contextmenutype: "native",
            tabIndex: 0,
            onChangeType: "onLeave"
        };
    });

    it("renders richtext widget", () => {
        const component = render(<RichText {...defaultProps} />);
        expect(component.container).toMatchSnapshot();
    });
});
