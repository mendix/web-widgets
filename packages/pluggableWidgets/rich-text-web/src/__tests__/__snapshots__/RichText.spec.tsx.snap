// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rich Text renders richtext widget 1`] = `
<div>
  <div
    class="widget-rich-text widget-rich-text-min-height"
    id="RichText1"
    style="width: 100%; min-height: 75px;"
  >
    <div
      class="mx-progress"
    />
  </div>
</div>
`;
